### PR TITLE
Create `auto_assign_transformers` and `get_transformers` for `Synthesizers`

### LINKED ISSUE #1020: Add assign_transformers and get_transformers methods to synthesizers
### Problem Description

_As a user, it would be useful if I could call a method that autonomously assigns transformers to my data using the metadata. It would also be helpful if I could examine those transformers afterwards._

#### Add `auto_assign_transformers` method to `BaseSynthesizer`
- Parameters:
    - data (`pd.DataFrame`): The raw data.
- This method should use the `DataProcessor` to do the following:
    1. Transform the constraints
    2. Create the config for the `HyperTransformer` and set it
- This method **SHOULD NOT** fit the `HyperTransformer`

#### Add `get_transformers` method to `BaseSynthesizer`
- This method should return a dictionary mapping the column name to the **actual transformer object** that will be used. This is essentially the `transformers` section of the `HyperTransformer`'s config.
- Errors:
    - If the `auto_assign_transformers` has not already been called, then this should crash
`Error: No transformers were returned in 'get_transformers'. Use 'assign_transformers' or 'fit' to create them.`

### Additional context

- The `auto_assign_transformers` method can be used in some circumstances in the `preprocess_data` method in #1018

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #1020

### CODE REVIEW

File: sdv/data_processing/data_processor.py [hunk 1]
```diff
@@ -314,6 +314,10 @@ def update_transformers(self, column_name_to_transformer):
 
         self._hyper_transformer.update_transformers(column_name_to_transformer)
 
+    def _create_hyper_transformer_instance(self):
```
Comment (amontanez24) [inline sdv/data_processing/data_processor.py:317]: why aren't we just using the `HyperTransformer` in the `DataProcessor`? We shouldn't need to make another one

File: sdv/single_table/base.py [hunk 2]
```diff
@@ -270,6 +270,44 @@ def get_metadata(self):
...
+
+        Args:
+            data (pandas.DataFrame):
+                The raw data (before any transformations) that will be used to fit the model.
+        """
+        constrained = self._data_processor._fit_transform_constraints(data)
+        columns_created_by_constraints = set(constrained.columns) - set(data.columns)
+        config = self._data_processor._create_config(constrained, columns_created_by_constraints)
+        self._data_processor._create_hyper_transformer_instance()
+        self._data_processor._hyper_transformer.set_config(config)
```
Comment (amontanez24) [inline sdv/single_table/base.py:287]: here we are missing that the formatters should also be fit. Can we instead make some methods in the `DataProcessor` that do these steps and then call them here?

**Review by amontanez24 — CHANGES_REQUESTED** [review]

File: sdv/single_table/base.py [hunk 3]
```diff
@@ -270,6 +270,44 @@ def get_metadata(self):
...
+        """Get a dictionary mapping of ``column_name``  and ``rdt.transformers``.
+
+        A dictionary representing the column names and the transformers that will be used
+        to transform the data.
+
+        Returns:
+            dict:
+                A dictionary mapping with column names and transformers.
+        """
+        field_transformers = {}
```
Comment (fealho) [inline sdv/single_table/base.py:299]: The logic can be tidied up a little:

```python3
if self._data_processor._hyper_transformer is not None:
    ft = self._data_processor._hyper_transformer.field_transformers
    if ft: return ft
raise
```

Comment (pvk-developer) [reply sdv/single_table/base.py:299]: Actually I forgot to change this logic, as `hyper_transformer` will no longer be `None`!

File: sdv/single_table/base.py [hunk 4]
```diff
@@ -270,6 +270,41 @@ def get_metadata(self):
...
+
+        Args:
+            data (pandas.DataFrame):
+                The raw data (before any transformations) that will be used to fit the model.
+        """
+        constrained = self._data_processor._fit_transform_constraints(data)
+        columns_created_by_constraints = set(constrained.columns) - set(data.columns)
+        self._data_processor._fit_numerical_formatters(data)
+        config = self._data_processor._create_config(constrained, columns_created_by_constraints)
+        self._data_processor._hyper_transformer.set_config(config)
```
Comment (amontanez24) [inline sdv/single_table/base.py:287]: These steps are very similar to what happens in the `DataProcessor.fit`. Should we make another method that sets up the hypertransformer as well as does the constraints and formatting in the `DataProcessor` and then call that method both here and in `DataProcessor.fit`?

File: sdv/data_processing/data_processor.py [hunk 5]
```diff
@@ -355,25 +349,35 @@ def _fit_numerical_formatters(self, data):
                 )
                 self.formatters[column_name].learn_format(data[column_name])
 
+    def _prepare_fitting(self, data):
```
Comment (amontanez24) [inline sdv/data_processing/data_processor.py:352]: we could make this method public honestly

**Review by amontanez24 — APPROVED** [review]
LGTM! Thanks for addressing the comments

File: sdv/data_processing/data_processor.py (hunk 5 above)
Comment (pvk-developer) [reply sdv/data_processing/data_processor.py]: Sounds good 👍🏻

**Review by fealho — APPROVED** [review]

**Comment by Grayangelz** [comment]
https://indico.math.cnrs.fr/

**Pull request merged**